### PR TITLE
Connection strings and docs

### PR DESCRIPTION
**Pull request by AsafMah**
-Re-did connection string to be more idiomatic and rustic. Will also help later when we add cloud info
-Wrote docs + examples (which are free tests in rust) for everything

### CODE REVIEW

File: azure-kusto-data/src/client.rs [hunk 1]
```diff
@@ -104,137 +116,170 @@ impl KustoClient {
...
+    /// assert!(matches!(result, KustoResponse::V1(..)));
+    /// # Ok(())}
+    /// ```
+    #[must_use]
+    pub fn execute_with_options(
         &self,
-        database: DB,
-        query: Q,
+        database: String,
+        query: String,
```
Comment (roeap) [inline azure-kusto-data/src/client.rs:145]: Personally I like trait based parameters - esecially when i started with rust there were a lot of conversions into strings etc going on ... `Into<String>` however is indeed maybe not the most convenient choice, but maybe we could do `AsRef<str>`. Lots of types implement that, and it's quite convenient from the usage perspective.

File: azure-kusto-data/src/client.rs [hunk 2]
```diff
@@ -104,137 +116,170 @@ impl KustoClient {
...
+    ///    let result = client.execute_command_with_options("some_database".into(), ".show version".into(),
+    ///     Some(RequestOptionsBuilder::default().with_request_app_name("app name".to_string()).build().unwrap()))
+    ///     .into_future().await?;
+    ///
+    /// for table in result.tables {
+    ///        println!("{}", table.table_name);
+    ///    }
+    /// # Ok(())}
+    /// ```
+    #[must_use]
```
Comment (roeap) [inline azure-kusto-data/src/client.rs:242]: Right now must_use probably makes sense here, but I was thinking abut the target state a bit ... if I remember correctly, we want to handle errors returned from kusto more explicitly, rather then having the user inspect the result tables for that.

If we had such error reporting already, would we still make this must use? With commands I suspect I often just want to know its executed without errors, and if so, an not necessarily interested in details - i.e. handle the erros, but the can be done without using the result?

File: azure-kusto-data/src/connection_string.rs [hunk 3]
```diff
@@ -135,267 +149,852 @@ static ALIAS_MAP: Lazy<HashMap<&'static str, ConnectionStringKey>> = Lazy::new(|
...
 
+        let mut result_map = HashMap::<ConnectionStringKey, &str>::new();
+
         for kv_pair_str in kv_str_pairs {
             let mut kv = kv_pair_str.trim().split('=');
             let k = match kv.next().filter(|k| !k.chars().all(char::is_whitespace)) {
                 None => {
                     return Err(ConnectionStringError::ParsingError {
-                        msg: "No key found".to_owned(),
+                        msg: "No key found".to_string(),
```
Comment (roeap) [inline azure-kusto-data/src/connection_string.rs:476]: one of the instances where he casting in rust gets a bit annoying 😆 . Maybe in a follow up PR we want to indocude some convenience functions I have seem in other crate like `ConnectionStringError::parsing_error(..)`.

File: azure-kusto-data/src/error.rs [hunk 4]
```diff
@@ -37,22 +39,39 @@ pub enum Error {
...
     #[error("Unexpected key '{}'", key)]
-    UnexpectedKey { key: String },
+    UnexpectedKey {
+        /// The key that is unexpected.
+        key: String,
+    },
+    /// Raised when a connection string has an invalid value.
     #[error("Parsing error: {}", msg)]
-    ParsingError { msg: String },
+    ParsingError {
```
Comment (roeap) [inline azure-kusto-data/src/error.rs:70]: should this just be `Parsing` to be consistent with other variant names?

File: azure-kusto-data/src/models.rs [hunk 5]
```diff
@@ -23,131 +25,220 @@ pub(crate) struct RequestProperties {
     pub(crate) options: Option<RequestOptions>,
 }
 
+/// Represents the scalar data types of ADX. see [the docs for more information](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/scalar-data-types/)
 #[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
-#[serde(rename_all = "snake_case")]
 pub enum ColumnType {
+    #[serde(alias = "Boolean", alias = "bool", alias = "boolean")]
+    /// Boolean type, true or false. Internally is a u8.
```
Comment (roeap) [inline azure-kusto-data/src/models.rs:32]: nice! much better than having many variants ❤️

**Review by roeap — COMMENTED** [review]
nice! I would say this bumps the user experience up by an order of magnitude 😄

File: azure-kusto-data/src/client.rs (hunk 2 above)
Comment (AsafMah) [reply azure-kusto-data/src/client.rs:242]: In the end, we would probably always want to return a future (right?), and a future does nothing until it's awaited or sent to the loop, so the user will always need to use it, even if we'll get errors before.

File: azure-kusto-data/src/client.rs (hunk 1 above)
Comment (AsafMah) [reply azure-kusto-data/src/client.rs]: I thought it makes it a bit too verbose, but then I remembered the impl Trait syntax and now it's very lean.
It does mean that some methods can't be const, but it's probably good enough for now.
WDYT?

File: azure-kusto-data/src/connection_string.rs (hunk 3 above)
Comment (AsafMah) [reply azure-kusto-data/src/connection_string.rs:478]: Did it, is that what you had in mind?

File: azure-kusto-data/src/error.rs (hunk 4 above)
Comment (AsafMah) [reply azure-kusto-data/src/error.rs]: Good call!

File: azure-kusto-data/src/client.rs (hunk 2 above)
Comment (roeap) [reply azure-kusto-data/src/client.rs:246]: I think I misunderstood what this means, and essentially thought you would have to assign it to something. But consuming it is using it :). The following would work as it consumes the future, right?

```rs
client.execute_command(...).await?;
```

File: azure-kusto-data/src/client.rs (hunk 1 above)
Comment (roeap) [reply azure-kusto-data/src/client.rs]: agreed!

**Review by roeap — APPROVED** [review]

**Comment by roeap** [comment]
Hi @AsafMah sorry for being MIA for so long, I think these changes look great and we can merge them :).

**Pull request merged**